x/ecocredit/server/testsuite/suite.go

```diff
@@ -618,45 +621,61 @@ func (s *IntegrationTestSuite) TestScenario() {
 		})
 	}
 
-	/****   TEST SET PRECISION   ****/
-	precisionCases := []struct {
-		name string
-		msg  ecocredit.MsgSetPrecision
-		ok   bool
+	/****   TEST CREDIT TYPES   ****/
+	creditTypeCases := []struct {
+		name        string
+		creditTypes []*ecocredit.CreditType
+		tx          ecocredit.MsgCreateClass
+		wantErr     bool
 	}{
 		{
-			"can NOT decrease the decimals", ecocredit.MsgSetPrecision{
-				Issuer: issuer1, BatchDenom: batchDenom, MaxDecimalPlaces: 2},
-			false,
-		}, {
-			"can NOT set to the same value", ecocredit.MsgSetPrecision{
-				Issuer: issuer1, BatchDenom: batchDenom, MaxDecimalPlaces: 7},
-			false,
-		}, {
-			"can increase", ecocredit.MsgSetPrecision{
-				Issuer: issuer1, BatchDenom: batchDenom, MaxDecimalPlaces: 8},
-			true,
-		}, {
-			"can NOT change precision of not existing denom", ecocredit.MsgSetPrecision{
-				Issuer: issuer1, BatchDenom: "not/existing", MaxDecimalPlaces: 1},
-			false,
+			name:        "valid eco credit creation",
+			creditTypes: []*ecocredit.CreditType{{Type: "carbon", Units: "kg", Precision: 3}},
```

Comment:
Shouldn't this fail, because the precision is wrong? If not, why not? Is our validation code for params actually called?

Reply:
Validation code only gets called when the param change route gets hit in the sdk code: [x/params/proposal_handler.go](https://github.com/cosmos/cosmos-sdk/blob/master/x/params/proposal_handler.go#L14-L24)

so here i am directly overriding the param space with the creditTypes field in the test case struct

Reply:
Cool, makes sense